# Lập trình cho khoa học dữ liệu: Đồ án cuối kỳ

**Danh sách thành viên:**

---
Họ và tên: La Ngọc Hồng Phúc

MSSV: 19127511

---
Họ và tên: Chung Thế Thọ

MSSV: 19127562

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Lập-trình-cho-khoa-học-dữ-liệu:-Đồ-án-cuối-kỳ" data-toc-modified-id="Lập-trình-cho-khoa-học-dữ-liệu:-Đồ-án-cuối-kỳ-1">Lập trình cho khoa học dữ liệu: Đồ án cuối kỳ</a></span><ul class="toc-item"><li><span><a href="#Môi-trường-code" data-toc-modified-id="Môi-trường-code-1.1">Môi trường code</a></span></li><li><span><a href="#Import-các-thư-viện-cần-thiết" data-toc-modified-id="Import-các-thư-viện-cần-thiết-1.2">Import các thư viện cần thiết</a></span></li><li><span><a href="#Thu-thập-dữ-liệu" data-toc-modified-id="Thu-thập-dữ-liệu-1.3">Thu thập dữ liệu</a></span></li><li><span><a href="#Khám-phá-dữ-liệu" data-toc-modified-id="Khám-phá-dữ-liệu-1.4">Khám phá dữ liệu</a></span><ul class="toc-item"><li><span><a href="#Giải-nén-dữ-liệu" data-toc-modified-id="Giải-nén-dữ-liệu-1.4.1">Giải nén dữ liệu</a></span></li><li><span><a href="#Đọc-dữ-liệu-từ-file-csv-vào-dataframe" data-toc-modified-id="Đọc-dữ-liệu-từ-file-csv-vào-dataframe-1.4.2">Đọc dữ liệu từ file csv vào dataframe</a></span><ul class="toc-item"><li><span><a href="#Tập-dữ-liệu-anime" data-toc-modified-id="Tập-dữ-liệu-anime-1.4.2.1">Tập dữ liệu anime</a></span></li><li><span><a href="#Tập-dữ-liệu-review" data-toc-modified-id="Tập-dữ-liệu-review-1.4.2.2">Tập dữ liệu review</a></span></li></ul></li><li><span><a href="#Mỗi-cột-có-ý-nghĩa-gì?" data-toc-modified-id="Mỗi-cột-có-ý-nghĩa-gì?-1.4.3">Mỗi cột có ý nghĩa gì?</a></span></li><li><span><a href="#Loại-bỏ-các-cột-không-dùng-đến" data-toc-modified-id="Loại-bỏ-các-cột-không-dùng-đến-1.4.4">Loại bỏ các cột không dùng đến</a></span><ul class="toc-item"><li><span><a href="#Tập-dữ-liệu-rating" data-toc-modified-id="Tập-dữ-liệu-rating-1.4.4.1">Tập dữ liệu rating</a></span></li></ul></li><li><span><a href="#Số-dòng,-số-cột-của-tập-dữ-liệu" data-toc-modified-id="Số-dòng,-số-cột-của-tập-dữ-liệu-1.4.5">Số dòng, số cột của tập dữ liệu</a></span><ul class="toc-item"><li><span><a href="#Tập-dữ-liệu-anime" data-toc-modified-id="Tập-dữ-liệu-anime-1.4.5.1">Tập dữ liệu anime</a></span></li><li><span><a href="#Tập-dữ-liệu-rating" data-toc-modified-id="Tập-dữ-liệu-rating-1.4.5.2">Tập dữ liệu rating</a></span></li><li><span><a href="#Nhận-xét:" data-toc-modified-id="Nhận-xét:-1.4.5.3">Nhận xét:</a></span></li></ul></li><li><span><a href="#Mỗi-dòng-có-ý-nghĩa-gì?-Có-vấn-đề-các-dòng-có-ý-nghĩa-khác-nhau-không?" data-toc-modified-id="Mỗi-dòng-có-ý-nghĩa-gì?-Có-vấn-đề-các-dòng-có-ý-nghĩa-khác-nhau-không?-1.4.6">Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?</a></span></li><li><span><a href="#Dữ-liệu-có-các-dòng-bị-lặp-không?" data-toc-modified-id="Dữ-liệu-có-các-dòng-bị-lặp-không?-1.4.7">Dữ liệu có các dòng bị lặp không?</a></span></li><li><span><a href="#Kiểu-dữ-liệu-của-từng-cột" data-toc-modified-id="Kiểu-dữ-liệu-của-từng-cột-1.4.8">Kiểu dữ liệu của từng cột</a></span><ul class="toc-item"><li><span><a href="#Tập-dữ-liệu-anime" data-toc-modified-id="Tập-dữ-liệu-anime-1.4.8.1">Tập dữ liệu anime</a></span></li><li><span><a href="#Tập-dữ-liệu-review" data-toc-modified-id="Tập-dữ-liệu-review-1.4.8.2">Tập dữ liệu review</a></span></li></ul></li><li><span><a href="#Phân-tích-rõ-hơn-các-cột-có-kiểu-dữ-liệu-object" data-toc-modified-id="Phân-tích-rõ-hơn-các-cột-có-kiểu-dữ-liệu-object-1.4.9">Phân tích rõ hơn các cột có kiểu dữ liệu object</a></span><ul class="toc-item"><li><span><a href="#Tập-dữ-liệu-anime" data-toc-modified-id="Tập-dữ-liệu-anime-1.4.9.1">Tập dữ liệu anime</a></span></li><li><span><a href="#Tập-dữ-liệu-review" data-toc-modified-id="Tập-dữ-liệu-review-1.4.9.2">Tập dữ liệu review</a></span></li></ul></li></ul></li><li><span><a href="#Tiền-xử-lý" data-toc-modified-id="Tiền-xử-lý-1.5">Tiền xử lý</a></span><ul class="toc-item"><li><span><a href="#Loại-bỏ-các-dòng-bị-lặp" data-toc-modified-id="Loại-bỏ-các-dòng-bị-lặp-1.5.1">Loại bỏ các dòng bị lặp</a></span></li><li><span><a href="#Chuyển-dtype-cột-genre-sang-list-of-string" data-toc-modified-id="Chuyển-dtype-cột-genre-sang-list-of-string-1.5.2">Chuyển dtype cột genre sang list of string</a></span></li><li><span><a href="#Chuyển-dtype-cột-aired-sang-dữ-liệu-thích-hợp" data-toc-modified-id="Chuyển-dtype-cột-aired-sang-dữ-liệu-thích-hợp-1.5.3">Chuyển dtype cột <code>aired</code> sang dữ liệu thích hợp</a></span></li></ul></li><li><span><a href="#Khám-phá-dữ-liệu-(tiếp-tục)" data-toc-modified-id="Khám-phá-dữ-liệu-(tiếp-tục)-1.6">Khám phá dữ liệu (tiếp tục)</a></span><ul class="toc-item"><li><span><a href="#Phân-bố-các-cột-dữ-liệu-dạng-numeric" data-toc-modified-id="Phân-bố-các-cột-dữ-liệu-dạng-numeric-1.6.1">Phân bố các cột dữ liệu dạng numeric</a></span><ul class="toc-item"><li><span><a href="#Tập-dữ-liệu-anime" data-toc-modified-id="Tập-dữ-liệu-anime-1.6.1.1">Tập dữ liệu anime</a></span></li><li><span><a href="#Tập-dữ-liệu-review" data-toc-modified-id="Tập-dữ-liệu-review-1.6.1.2">Tập dữ liệu review</a></span></li></ul></li><li><span><a href="#Phân-bố-các-cột-dữ-liệu-dạng-categorical" data-toc-modified-id="Phân-bố-các-cột-dữ-liệu-dạng-categorical-1.6.2">Phân bố các cột dữ liệu dạng categorical</a></span><ul class="toc-item"><li><span><a href="#Tập-dữ-liệu-anime" data-toc-modified-id="Tập-dữ-liệu-anime-1.6.2.1">Tập dữ liệu anime</a></span></li><li><span><a href="#Tập-dữ-liệu-review" data-toc-modified-id="Tập-dữ-liệu-review-1.6.2.2">Tập dữ liệu review</a></span></li></ul></li><li><span><a href="#Khám-phá-thêm-(nếu-có)" data-toc-modified-id="Khám-phá-thêm-(nếu-có)-1.6.3">Khám phá thêm (nếu có)</a></span></li></ul></li><li><span><a href="#Đưa-ra-các-câu-hỏi-cần-trả-lời" data-toc-modified-id="Đưa-ra-các-câu-hỏi-cần-trả-lời-1.7">Đưa ra các câu hỏi cần trả lời</a></span></li></ul></li></ul></div>

## Môi trường code

In [30]:
import sys
sys.executable

'/home/stazjac/anaconda3/envs/min_ds-env/bin/python'

## Import các thư viện cần thiết

In [226]:
# Import thư viện cần dùng
import os
import ast
import zipfile
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Thu thập dữ liệu

Trong đồ án này, ta sẽ sử dụng tập dữ liệu có sẵn đến từ 2 file mà người ta đã thu thập trên kaggle là `animes.csv` và `reviews.csv`. Theo như mô tả của người thu thập dữ liệu thì ta có khoảng 16000 anime và 130000 bài review về các anime này. Cách thức thu thập dữ liệu của người này là dùng API do trang myanimelist.net cung cấp, cụ thể chi tiết về chương trình crawler mà anh ta viết có thể được xem ở link này:  https://github.com/marlesson/scrapy_myanimelist

Link của tập dữ liệu: [Anime Dataset with Reviews - MyAnimeList](https://www.kaggle.com/marlesson/myanimelist-dataset-animes-profiles-reviews) ([https://www.kaggle.com/marlesson/myanimelist-dataset-animes-profiles-reviews](https://www.kaggle.com/marlesson/myanimelist-dataset-animes-profiles-reviews))

---

## Khám phá dữ liệu

### Giải nén dữ liệu
Sau khi tải tập dữ liệu về, chúng ta cần giải nén và tiền xử lý chúng

In [32]:
ZIP_FILE = 'archive.zip'
DATA_DIR = os.path.join(os.getcwd(), 'data')
ANIME_CSV = os.path.join(DATA_DIR, 'animes.csv')
REVIEW_CSV = os.path.join(DATA_DIR, 'reviews.csv')

OVERWRITE = False

def extractall():
    with zipfile.ZipFile(ZIP_FILE, 'r') as zip_ref:
        zip_ref.extractall(DATA_DIR)

def file_sizes(pathdir):
    join = os.path.join
    getsize = os.path.getsize
    isfile = lambda f: os.path.isfile(join(pathdir, f))
    files = filter(isfile, os.listdir(pathdir))
    fsizes = [(f, getsize(join(pathdir, f)) / 1024 ** 2) for f in files]
    return pd.DataFrame(fsizes, columns=['File Name', 'Size (MB)'])

if OVERWRITE or not os.path.isdir(DATA_DIR) or not os.listdir(DATA_DIR):
    extractall()

file_sizes(DATA_DIR)

File Name   Size (MB)
0   reviews.csv  654.141182
1    animes.csv   11.514027
2  profiles.csv    8.262743

### Đọc dữ liệu từ file csv vào dataframe

#### Tập dữ liệu anime

In [236]:
anime_df = pd.read_csv(ANIME_CSV)
anime_df.head()

uid                             title  \
0  28891           Haikyuu!! Second Season   
1  23273           Shigatsu wa Kimi no Uso   
2  34599                     Made in Abyss   
3   5114  Fullmetal Alchemist: Brotherhood   
4  31758  Kizumonogatari III: Reiketsu-hen   

                                            synopsis  \
0  Following their participation at the Inter-Hig...   
1  Music accompanies the path of the human metron...   
2  The Abyss—a gaping chasm stretching down into ...   
3  "In order for something to be obtained, someth...   
4  After helping revive the legendary vampire Kis...   

                                               genre  \
0  ['Comedy', 'Sports', 'Drama', 'School', 'Shoun...   
1  ['Drama', 'Music', 'Romance', 'School', 'Shoun...   
2  ['Sci-Fi', 'Adventure', 'Mystery', 'Drama', 'F...   
3  ['Action', 'Military', 'Adventure', 'Comedy', ...   
4   ['Action', 'Mystery', 'Supernatural', 'Vampire']   

                          aired  episodes  members  popularity  ranked  score  \
0   Oct 4, 2015 to Mar 27, 2016        25   489888         141      25      9   
1  Oct 10, 2014 to Mar 20, 2015        22   995473          28      24      9   
2   Jul 7, 2017 to Sep 29, 2017        13   581663          98      23      9   
3    Apr 5, 2009 to Jul 4, 2010        64  1615084           4       1      9   
4                   Jan 6, 2017         1   214621         502      22      9   

                                             img_url  \
0  https://cdn.myanimelist.net/images/anime/9/766...   
1  https://cdn.myanimelist.net/images/anime/3/671...   
2  https://cdn.myanimelist.net/images/anime/6/867...   
3  https://cdn.myanimelist.net/images/anime/1223/...   
4  https://cdn.myanimelist.net/images/anime/3/815...   

                                                link  
0  https://myanimelist.net/anime/28891/Haikyuu_Se...  
1  https://myanimelist.net/anime/23273/Shigatsu_w...  
2  https://myanimelist.net/anime/34599/Made_in_Abyss  
3  https://myanimelist.net/anime/5114/Fullmetal_A...  
4  https://myanimelist.net/anime/31758/Kizumonoga...

#### Tập dữ liệu review

Giải nén xong thì mình thấy file `reviews.csv` tới gần 700MB, nên mình sẽ không load trực tiếp mà sẽ xem trước có cái nào có thể lược bỏ để giảm bớt dung lượng

In [34]:
pd.read_csv(REVIEW_CSV, nrows=5)

uid          profile  anime_uid  \
0  255938   DesolatePsyche      34096   
1  259117        baekbeans      34599   
2  253664             skrn      28891   
3    8254     edgewalker00       2904   
4  291149  aManOfCulture99       4181   

                                                text  score  \
0  \n           \n         \n           \n       ...      8   
1  \n           \n         \n           \n       ...     10   
2  \n           \n         \n           \n       ...      7   
3  \n           \n         \n           \n       ...      9   
4  \n           \n         \n           \n       ...     10   

                                              scores  \
0  {'Overall': '8', 'Story': '8', 'Animation': '8...   
1  {'Overall': '10', 'Story': '10', 'Animation': ...   
2  {'Overall': '7', 'Story': '7', 'Animation': '9...   
3  {'Overall': '9', 'Story': '9', 'Animation': '9...   
4  {'Overall': '10', 'Story': '10', 'Animation': ...   

                                            link  
0  https://myanimelist.net/reviews.php?id=255938  
1  https://myanimelist.net/reviews.php?id=259117  
2  https://myanimelist.net/reviews.php?id=253664  
3    https://myanimelist.net/reviews.php?id=8254  
4  https://myanimelist.net/reviews.php?id=291149

### Mỗi cột có ý nghĩa gì?

Mô tả trên kaggle về các cột trong file `animes.csv`:
- **uid**: ID của anime
- **title**: tiêu đề của anime
- **synopsis**: giới thiệu nội dung của anime
- **genre**: thể loại của anime
- **aired**: ngày phát hành anime và ngày kết thúc của nó (nếu có)
- **episodes**: tổng số tập của anime
- **members**: tổng số thành viên trong cộng đồng quan tâm anime này
- **popularity**: độ "nổi tiếng" của anime, số càng thấp thì càng nổi tiếng
- **ranked**: xếp hạng của anime, số càng thấp hạng càng cao và anime đó càng hay
- **score**: điểm đánh giá anime trên thang điểm 10 từ trang myanimelist
- **img_url**: link ảnh bìa của anime
- **link**: link đến anime trên myanimelist

Mô tả trên kaggle về các cột trong file `reviews.csv`:
- **uid**: ID của bài review
- **profile**: tên của user thực hiện review
- **anime_uid**: ID của anime được review
- **text**: đoạn văn nội dung mà user đó review
- **score**: điểm tổng quát user này đã đánh giá cho anime
- **scores**: điểm chi tiết từng phần mà user này đã đánh giá cho anime bao gồm
    - Overall: điểm tổng
    - Story: cốt truyện
    - Animation: đồ họa, nét vẽ
    - Sound: âm thanh
    - Character: xây dựng nhân vật
    - Enjoyment: độ giải trí
- **link**: đường dẫn URL đến bài review này

### Loại bỏ các cột không dùng đến

Trong phạm vi môn học, mình sẽ bỏ qua các cột cần dùng đến kỹ thuật xử lý ngôn ngữ tự nhiên cũng như các cột liên quan đến URL. Việc bỏ các cột này giúp ta chạy code nhanh hơn vì các cột này rất nặng. Như vậy đối với mỗi tập dữ liệu sẽ bỏ các cột sau:

- **anime**: synopsis, img_url, link
- **review**: text, link  

Đối với **review** thì do file khá lớn nên mình sẽ load theo từng chunk, loại bỏ các cột không cần thiết và lưu nó vào file `ratings.csv`

In [36]:
# Loại bỏ cột text và link trong review
RATING_CSV = os.path.join(DATA_DIR, 'ratings.csv')

def review_to_rating():
    is_first = True
    for chunk in pd.read_csv(REVIEW_CSV, chunksize=1000):
        chunk.drop(columns=['text', 'link'], inplace=True)
        chunk.to_csv(RATING_CSV, mode='a', index=False, header=is_first)
        is_first = False

if not os.path.isfile(RATING_CSV):
    review_to_rating()

print("File name: ratings.csv")
print("Size (MB):", os.path.getsize(RATING_CSV) / 1024 ** 2)

File name: ratings.csv
Size (MB): 23.345977783203125


#### Tập dữ liệu rating

In [37]:
rating_df = pd.read_csv(RATING_CSV)
rating_df.head()

uid          profile  anime_uid  score  \
0  255938   DesolatePsyche      34096      8   
1  259117        baekbeans      34599     10   
2  253664             skrn      28891      7   
3    8254     edgewalker00       2904      9   
4  291149  aManOfCulture99       4181     10   

                                              scores  
0  {'Overall': '8', 'Story': '8', 'Animation': '8...  
1  {'Overall': '10', 'Story': '10', 'Animation': ...  
2  {'Overall': '7', 'Story': '7', 'Animation': '9...  
3  {'Overall': '9', 'Story': '9', 'Animation': '9...  
4  {'Overall': '10', 'Story': '10', 'Animation': ...

So với file `reviews.csv` thì file `ratings.csv` nhỏ hơn rất nhiều, có thể load vào DataFrame mà không cần chia chunk nữa

In [38]:
# Loại bỏ cột synopsis, img_url, link trong anime
anime_df.drop(columns=['synopsis', 'img_url', 'link'], inplace=True)
anime_df.head()

uid                             title  \
0  28891           Haikyuu!! Second Season   
1  23273           Shigatsu wa Kimi no Uso   
2  34599                     Made in Abyss   
3   5114  Fullmetal Alchemist: Brotherhood   
4  31758  Kizumonogatari III: Reiketsu-hen   

                                               genre  \
0  ['Comedy', 'Sports', 'Drama', 'School', 'Shoun...   
1  ['Drama', 'Music', 'Romance', 'School', 'Shoun...   
2  ['Sci-Fi', 'Adventure', 'Mystery', 'Drama', 'F...   
3  ['Action', 'Military', 'Adventure', 'Comedy', ...   
4   ['Action', 'Mystery', 'Supernatural', 'Vampire']   

                          aired  episodes  members  popularity  ranked  score  
0   Oct 4, 2015 to Mar 27, 2016      25.0   489888         141    25.0   8.82  
1  Oct 10, 2014 to Mar 20, 2015      22.0   995473          28    24.0   8.83  
2   Jul 7, 2017 to Sep 29, 2017      13.0   581663          98    23.0   8.83  
3    Apr 5, 2009 to Jul 4, 2010      64.0  1615084           4     1.0   9.23  
4                   Jan 6, 2017       1.0   214621         502    22.0   8.83

### Số dòng, số cột của tập dữ liệu

#### Tập dữ liệu anime

In [39]:
anime_num_rows, anime_num_cols = anime_df.shape
print('Số dòng:', anime_num_rows)
print('Số cột:', anime_num_cols)

Số dòng: 19311
Số cột: 9


#### Tập dữ liệu rating

In [40]:
rating_num_rows, rating_num_cols = rating_df.shape
print('Số dòng:', rating_num_rows)
print('Số cột:', rating_num_cols)

Số dòng: 192112
Số cột: 5


#### Nhận xét:

Có vẻ số dòng từ cả 2 tập dữ liệu `anime_df` và `rating_df` trả về khá khác so với mô tả:
- `anime_df`: 19000 so với 16000
- `rating_df`: 190000 so với 130000.  

**Lưu ý**: tập dữ liệu `ratings.csv` không khác nhau về số dòng so với tập dữ liệu `reviews.csv`

In [41]:
# Chứng minh số dòng của rating.csv không khác so với review.csv
import csv
with open(REVIEW_CSV, 'r') as f:
    csvreader = csv.reader(f)
    next(csvreader) # Bỏ qua header
    assert rating_num_rows == sum(1 for row in csvreader)

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

- Tập dữ liệu anime: mỗi dòng trong tập này là thông tin của một anime viết bởi trang myanimelist.net
- Tập dữ liệu review: mỗi dòng trong tập này là thông tin về review của một user dành cho một anime trên trang myanimelist.net

Có vẻ như mỗi dòng trong các tập không có gì bất ổn.

### Dữ liệu có các dòng bị lặp không?

In [42]:
anime_has_dup_rows = np.sum(anime_df.duplicated())
rating_has_dup_rows = np.sum(rating_df.duplicated())

pd.DataFrame({'Dataset': ['Anime', 'Rating'],
             'Duplicate count': [anime_has_dup_rows,
             rating_has_dup_rows]})

Dataset  Duplicate count
0   Anime             2943
1  Rating            61593

### Kiểu dữ liệu của từng cột

#### Tập dữ liệu anime

In [43]:
# Xem kiểu dữ liệu của từng cột tập anime
anime_df.dtypes

uid             int64
title          object
genre          object
aired          object
episodes      float64
members         int64
popularity      int64
ranked        float64
score         float64
dtype: object

#### Tập dữ liệu review

In [44]:
# Xem kiểu dữ liệu của từng cột tập review
rating_df.dtypes

uid           int64
profile      object
anime_uid     int64
score         int64
scores       object
dtype: object

### Phân tích rõ hơn các cột có kiểu dữ liệu object

#### Tập dữ liệu anime

In [45]:
def open_object_dtype(s):
    return set(s.apply(type))
    
print('title:', open_object_dtype(anime_df['title']))
print('genre:', open_object_dtype(anime_df['genre']))
print('aired:', open_object_dtype(anime_df['aired']))

title: {<class 'str'>}
genre: {<class 'str'>}
aired: {<class 'str'>}


#### Tập dữ liệu review

In [46]:
print('profile:', open_object_dtype(review_df['profile']))
print('scores:', open_object_dtype(review_df['scores']))

profile: {<class 'str'>}
scores: {<class 'str'>}


**Các vấn đề cần tiền xử lý:**
- Loại bỏ các dòng bị lặp
- Chuyển dtype cột `genre` sang list of string
- Chuyển dtype cột `aired` sang loại dữ liệu thích hợp

---

## Tiền xử lý

### Loại bỏ các dòng bị lặp

In [237]:
# Loại bỏ các dòng bị duplicate
anime_df = anime_df.drop_duplicates()
rating_df = rating_df.drop_duplicates()

In [238]:
# In lại số dòng bị lặp sau khi loại bỏ duplicate để check
anime_has_dup_rows = np.sum(anime_df.duplicated())
rating_has_dup_rows = np.sum(rating_df.duplicated())

pd.DataFrame({'Dataset': ['Anime', 'Rating'],
             'Duplicate count': [anime_has_dup_rows,
             rating_has_dup_rows]})

Dataset  Duplicate count
0   Anime                0
1  Rating                0

### Chuyển dtype cột genre sang list of string

In [239]:
anime_df['genre'] = anime_df.apply({'genre':ast.literal_eval})
anime_df.head()

uid                             title  \
0  28891           Haikyuu!! Second Season   
1  23273           Shigatsu wa Kimi no Uso   
2  34599                     Made in Abyss   
3   5114  Fullmetal Alchemist: Brotherhood   
4  31758  Kizumonogatari III: Reiketsu-hen   

                                            synopsis  \
0  Following their participation at the Inter-Hig...   
1  Music accompanies the path of the human metron...   
2  The Abyss—a gaping chasm stretching down into ...   
3  "In order for something to be obtained, someth...   
4  After helping revive the legendary vampire Kis...   

                                               genre  \
0           [Comedy, Sports, Drama, School, Shounen]   
1           [Drama, Music, Romance, School, Shounen]   
2       [Sci-Fi, Adventure, Mystery, Drama, Fantasy]   
3  [Action, Military, Adventure, Comedy, Drama, M...   
4           [Action, Mystery, Supernatural, Vampire]   

                          aired  episodes  members  popularity  ranked  score  \
0   Oct 4, 2015 to Mar 27, 2016        25   489888         141      25      9   
1  Oct 10, 2014 to Mar 20, 2015        22   995473          28      24      9   
2   Jul 7, 2017 to Sep 29, 2017        13   581663          98      23      9   
3    Apr 5, 2009 to Jul 4, 2010        64  1615084           4       1      9   
4                   Jan 6, 2017         1   214621         502      22      9   

                                             img_url  \
0  https://cdn.myanimelist.net/images/anime/9/766...   
1  https://cdn.myanimelist.net/images/anime/3/671...   
2  https://cdn.myanimelist.net/images/anime/6/867...   
3  https://cdn.myanimelist.net/images/anime/1223/...   
4  https://cdn.myanimelist.net/images/anime/3/815...   

                                                link  
0  https://myanimelist.net/anime/28891/Haikyuu_Se...  
1  https://myanimelist.net/anime/23273/Shigatsu_w...  
2  https://myanimelist.net/anime/34599/Made_in_Abyss  
3  https://myanimelist.net/anime/5114/Fullmetal_A...  
4  https://myanimelist.net/anime/31758/Kizumonoga...

### Chuyển dtype cột `aired` sang dữ liệu thích hợp

**Xem format của cột này trong tập dữ liệu**

In [240]:
anime_df['aired'].sample(10)

11659    Oct 11, 1980 to Jun 13, 1982
13953                            1990
1891     Aug 27, 2018 to Aug 31, 2018
11357                    May 25, 2007
6142                     May 30, 2012
16228                    Apr 25, 1998
12699                            2010
3995                     Sep 14, 2018
2844                     Jul 18, 2001
2057     Jan 21, 1987 to Apr 10, 1989
Name: aired, dtype: object

Sau khi lấy 10 mẫu ngẫu nhiên của cột này khoảng 10 lần thì có thể rút ra như sau:
- Định dạng trong pandas cột sử dụng là `%b %d, %Y`
- Có mẫu có cả ngày bắt đầu và kết thúc, có mẫu chỉ có ngày bắt đầu
- Đa phần các mẫu đều đầy đủ ngày tháng năm, tuy nhiên có vài mẫu chỉ có tháng và năm, thậm chỉ chỉ có năm.

Trong phạm vi đồ án, ta chỉ quan tâm đến thời điểm anime đó bắt đầu. Như vậy,đầu tiên ta sẽ trích xuất thời điểm bắt đầu trước rồi sau đó chuyển sang `datetime` theo định dạng `%b %d, %Y` thử, các giá trị thiếu ngày hoặc thiếu tháng ta sẽ tạm để NaN và đếm xem có bao nhiêu mẫu có giá trị như vậy.

In [241]:
# Dùng regular expresion trích xuất ngày bắt đầu
start_date = anime_df['aired'].str.extract(r"^([a-zA-Z]{3}\s\d{1,2},\s\d{4})",
                                          expand=False #Trả về Series
                                          )

start_date = pd.to_datetime(start_date,
                           format='%b %d, %Y',
                           errors='ignore' # Bỏ qua NaN
                           )

start_date

0       2015-10-04
1       2014-10-10
2       2017-07-07
3       2009-04-05
4       2017-01-06
           ...    
19002   2011-01-01
19003   2018-01-12
19004   2011-09-27
19005   2016-07-09
19006   2013-06-19
Name: aired, Length: 16368, dtype: datetime64[ns]

In [242]:
# Đếm số dòng không đầy đủ ngày hoặc tháng
start_date.isnull().sum()

1990

Để đảm bảo nhận định việc cột `aired` có các mẫu thiếu ngày hoặc tháng, ta thử xem lại các dòng NaN trong **start_date** thì trong tập dữ liệu `anime_df`, các dòng này sẽ có dữ liệu như thế nào.

In [251]:
anime_df[start_date.isnull()]

uid                                              title  \
327     8740             One Piece Film: Strong World Episode 0   
406    26117                                           Kokoroya   
518    30298  Mahou no Princess Minky Momo: Yume wo Dakishim...   
524    35628      Honoo no Alpenrose: Ai no Symphony Ongaku-hen   
795     6154  Saint Seiya: Meiou Hades Juuni Kyuu-hen - Yomi...   
...      ...                                                ...   
18090  10334                              Rurouni Kenshin Recap   
18160   5947  Tengen Toppa Gurren Lagann Movie Zenyasai: Vir...   
18635   2512                Detective Conan OVA 02: 16 Suspects   
18663   2513  Detective Conan OVA 03: Conan and Heiji and th...   
18946  33335            Hajimete no Eigo: Nippon Mukashibanashi   

                                                synopsis  \
327    Set over 20 years prior to the main One Piece ...   
406    An educational film about empathy in Japan. In...   
518                              Three unaired episodes.   
524    A music-video style recap OVA. This release co...   
795    Is a summary of the Saga Hades Sanctuary from ...   
...                                                  ...   
18090               This is a summary of episodes 34-56.   
18160  Viral tries to get some sleep, but keeps on ha...   
18635  Conan Edogawa and the rest of the gang are inv...   
18663  Conan and Ran are visiting Heiji and Kazuha in...   
18946  Old Japanese folk legends released on Japanese...   

                                                   genre          aired  \
327    [Action, Adventure, Comedy, Fantasy, Shounen, ...      Apr, 2010   
406                        [Kids, Psychological, Sci-Fi]           2014   
518                                      [Magic, Shoujo]  Not available   
524          [Drama, Historical, Music, Romance, Shoujo]  Not available   
795                         [Sci-Fi, Adventure, Shounen]           2003   
...                                                  ...            ...   
18090  [Action, Adventure, Comedy, Historical, Romanc...      Jul, 1997   
18160                                           [Comedy]      Jan, 2009   
18635      [Adventure, Mystery, Comedy, Police, Shounen]           2002   
18663      [Adventure, Mystery, Comedy, Police, Shounen]           2003   
18946                                    [Kids, Fantasy]  Not available   

       episodes  members  popularity  ranked  score  \
327           1    54305        1793     475      8   
406           1      366       11749     394      8   
518           3      398       11591     281      8   
524           1      314       12128     274      8   
795           2     7675        4829     848      8   
...         ...      ...         ...     ...    ...   
18090         1     6600        5135    2381      7   
18160         1    24078        2909    2307      7   
18635         1    12089        4002    2182      7   
18663         1    11463        4101    2695      7   
18946        18      389       11599    2960      7   

                                                 img_url  \
327    https://cdn.myanimelist.net/images/anime/2/241...   
406    https://cdn.myanimelist.net/images/anime/2/659...   
518    https://cdn.myanimelist.net/images/anime/5/726...   
524    https://cdn.myanimelist.net/images/anime/11/85...   
795    https://cdn.myanimelist.net/images/anime/7/135...   
...                                                  ...   
18090  https://cdn.myanimelist.net/images/anime/13/28...   
18160  https://cdn.myanimelist.net/images/anime/13/22...   
18635  https://cdn.myanimelist.net/images/anime/10/71...   
18663  https://cdn.myanimelist.net/images/anime/13/72...   
18946  https://cdn.myanimelist.net/images/anime/6/800...   

                                                    link  year_aired  
327    https://myanimelist.net/anime/8740/One_Piece_F...        2010  
406         https://myanimelist.net/anime/26117/Kokoroya  

Qua kết quả trên ta còn thấy có các mẫu không xác định được ngày phát hành "Not available". Con số 1990 mẫu thiếu trên khoảng 16000 mẫu cũng là đáng kể. Vì vậy để giảm thiểu con số này, nhóm mình quyết định chỉ lấy năm phát hành của anime. Điều này có vẻ là chấp nhận được vì dữ liệu của ta trải dài hơn 20 năm và mỗi lần có các phân tích đến "trending" thì ta có thể làm trên năm thay vì theo tháng ngày. Năm phát hành của các anime sẽ thành cột mới **year_aired** trong tập dữ liệu `anime_df` phòng trường hợp sau này ta có các phân tích cần chính xác đến ngày và tháng.

In [244]:
# Dùng regular expresion trích xuất năm bắt đầu
year_aired = anime_df['aired'].str.extract(r"(\d{4})",
                                          expand=False #Trả về Series
                                          )
year_aired = pd.to_numeric(year_aired)
year_aired

0       2015
1       2014
2       2017
3       2009
4       2017
        ... 
19002   2011
19003   2018
19004   2011
19005   2016
19006   2013
Name: aired, Length: 16368, dtype: float64

In [245]:
# Đếm số dòng không có ngày phát hành
year_aired.isnull().sum()

295

Số dòng NaN giờ đây chỉ còn 295, giảm đi đáng kể! Tiếp theo ta thêm cột năm này vào `anime_df`.

In [246]:
anime_df['year_aired'] = year_aired
anime_df

uid                                              title  \
0      28891                            Haikyuu!! Second Season   
1      23273                            Shigatsu wa Kimi no Uso   
2      34599                                      Made in Abyss   
3       5114                   Fullmetal Alchemist: Brotherhood   
4      31758                   Kizumonogatari III: Reiketsu-hen   
...      ...                                                ...   
19002  10075                                        Naruto x UT   
19003  35828                                   Miira no Kaikata   
19004  10378                             Shinryaku!? Ika Musume   
19005  33082                      Kingsglaive: Final Fantasy XV   
19006  16934  Chuunibyou demo Koi ga Shitai!: Kirameki no......   

                                                synopsis  \
0      Following their participation at the Inter-Hig...   
1      Music accompanies the path of the human metron...   
2      The Abyss—a gaping chasm stretching down into ...   
3      "In order for something to be obtained, someth...   
4      After helping revive the legendary vampire Kis...   
...                                                  ...   
19002  All-new animation offered throughout UNIQLO cl...   
19003  High school student Sora Kashiwagi is accustom...   
19004  After regaining her squid-like abilities, Ika ...   
19005  For years, the Niflheim Empire and the kingdom...   
19006  Although Yuuta Togashi and Rikka Takanashi hav...   

                                                   genre  \
0               [Comedy, Sports, Drama, School, Shounen]   
1               [Drama, Music, Romance, School, Shounen]   
2           [Sci-Fi, Adventure, Mystery, Drama, Fantasy]   
3      [Action, Military, Adventure, Comedy, Drama, M...   
4               [Action, Mystery, Supernatural, Vampire]   
...                                                  ...   
19002  [Action, Comedy, Super Power, Martial Arts, Sh...   
19003              [Slice of Life, Comedy, Supernatural]   
19004                   [Slice of Life, Comedy, Shounen]   
19005                                           [Action]   
19006    [Comedy, Drama, Romance, School, Slice of Life]   

                              aired  episodes  members  popularity  ranked  \
0       Oct 4, 2015 to Mar 27, 2016        25   489888         141      25   
1      Oct 10, 2014 to Mar 20, 2015        22   995473          28      24   
2       Jul 7, 2017 to Sep 29, 2017        13   581663          98      23   
3        Apr 5, 2009 to Jul 4, 2010        64  1615084           4       1   
4                       Jan 6, 2017         1   214621         502      22   
...                             ...       ...      ...         ...     ...   
19002                   Jan 1, 2011         1    34155        2382    1728   
19003  Jan 12, 2018 to Mar 30, 2018        12    61459        1648    1727   
19004  Sep 27, 2011 to Dec 25, 2011        12    67422        1547    1548   
19005                   Jul 9, 2016         1    41077        2154    1544   
19006                  Jun 19, 2013         1   128400         883    1533   

       score                                            img_url  \
0          9  https://cdn.myanimelist.net/images/anime/9/766...   
1          9  https://cdn.myanimelist.net/images/anime/3/671...   
2          9  https://cdn.myanimelist.net/images/anime/6/867...   
3          9  https://cdn.myanimelist.net/images/anime/1223/...   
4          9  https://cdn.myanimelist.net/images/anime/3/815...   
...      ...                                                ...   
19002      8  https://cdn.myanimelist.net/images/anime/3/304...   
19003      8  https://cdn.myanimelist.net/images/anime/1486/...   
19004      8  https://cdn.myanimelist.net/images/anime/6/301...   
19005      8  https://cdn.myanimelist.net/images/anime/12/79...   
19006      8  https://cdn.myanimelist.net/images/anime/7/455...   

                    

---

## Khám phá dữ liệu (tiếp tục)

### Phân bố các cột dữ liệu dạng numeric

Ở mỗi cột ta sẽ xem xét % tỉ lệ thiếu dữ liệu (0 đến 100), giá trị nhỏ nhất và lớn nhất.

#### Tập dữ liệu anime

Các cột numeric: episodes, members, popularity, ranked, score

In [247]:
pd.options.display.float_format = '{:.0f}'.format

def missing_ratio(s):
    return s.isna().mean() * 100

def min_value(s):
    return s.dropna().min()

def max_value(s):
    return s.dropna().max()

anime_numeric_cols = ['episodes', 'members', 'popularity', 'ranked', 'score', 'year_aired']
anime_nume_col_profiles_df = anime_df[anime_numeric_cols]

anime_nume_col_profiles_df = anime_nume_col_profiles_df.agg([missing_ratio,
                                                             min_value,
                                                             max_value])

anime_nume_col_profiles_df

episodes  members  popularity  ranked  score  year_aired
missing_ratio         3        0           0      11      2           2
min_value             1       25           1       1      1        1917
max_value          3057  1871043       16338   14675      9        2021

#### Tập dữ liệu review

Các cột numeric: score

In [248]:
review_nume_col_profiles_df = review_df['score']

review_nume_col_profiles_df = review_nume_col_profiles_df.agg([missing_ratio,
                                                             min_value,
                                                             max_value])

review_nume_col_profiles_df

missing_ratio    0
min_value        7
max_value       10
Name: score, dtype: float64

### Phân bố các cột dữ liệu dạng categorical

Ở mỗi cột ta sẽ xem xét % tỉ lệ thiếu dữ liệu (0 đến 100), số giá trị khác nhau (không xét giá trị thiếu)

#### Tập dữ liệu anime

Các cột categorical: uid, title, genre. Ta không xét cột **aired** vì nó là cột dữ liệu thô. Hai cột đầu thì ta có thể xử lý nhanh bằng method agg với các hàm tự định nghĩa, riêng cột `genre` ta cần phải đi đếm số thể loại khác nhau vì cột này lưu dưới dạng list of string. Để đếm số thể loại khác nhau, ta có thể hợp các list này lại rồi đếm số giá trị.



In [286]:
anime_cate_cols = ['uid', 'title']
anime_cate_col_profiles_df = anime_df[anime_cate_cols]

def num_diff_vals(s):
    return s.dropna().nunique()

anime_cate_col_profiles_df = anime_cate_col_profiles_df.agg([missing_ratio,
                                                             num_diff_vals])

genre_profile = [missing_ratio(anime_df['genre']),
                 anime_df['genre'].apply(pd.Series).stack().nunique()]

genre_profile = pd.DataFrame(genre_profile,
                             index=['missing_ratio', 'num_diff_vals'],
                             columns=['genre'])

anime_cate_col_profiles_df = pd.concat([anime_cate_col_profiles_df, genre_profile],
                                       axis=1)

anime_cate_col_profiles_df

uid  title  genre
missing_ratio     0      0      0
num_diff_vals 16216  16214     43

#### Tập dữ liệu review

Các cột categorical: uid, profile, anime_uid

In [287]:
rating_cate_cols = ['uid', 'profile', 'anime_uid']
rating_cate_col_profiles_df = rating_df[rating_cate_cols]

def num_diff_vals(s):
    return s.dropna().nunique()

rating_cate_col_profiles_df = rating_cate_col_profiles_df.agg([missing_ratio,
                                                             num_diff_vals])
rating_cate_col_profiles_df

uid  profile  anime_uid
missing_ratio      0        0          0
num_diff_vals 130519    47885       8113

---

## Đưa ra các câu hỏi cần trả lời